auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,198 @@
+contract;
+
+dep abi;
+dep errors;
+dep data_structures;
+dep events;
+
+use std::{
+    address::Address,
+    assert::require,
+    block::height,
+    chain::auth::{AuthError, msg_sender},
+    context::{call_frames::msg_asset_id, msg_amount},
+    contract_id::ContractId,
+    identity::Identity,
+    logging::log,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::{force_transfer_to_contract, transfer_to_output}
+};
+
+use abi::DutchAuction;
+use errors::Error;
+use data_structures::Auction;
+use events::*;
+
+storage {
+    /// Whether or not the constructor function has been called yet
+    initialized: bool,
+    /// Mapping an auction_id to its respective auction, allowing for multiple auctions to happen simultaneously
+    auctions: StorageMap<u64,
+    Auction>, /// The Admin Address
+    admin: Identity,
+    /// Tracking how many auctions have been made till now
+    auction_count: u64,
+}
+
+impl DutchAuction for Contract {
+    fn constructor(admin: Identity) {
+        require(!storage.initialized, Error::CannotReinitialize);
+        storage.admin = admin;
+        storage.initialized = true;
+    }
+
+    fn price(auction_id: u64) -> u64 {
+        /// If the given auction id is higher than the auction count, its an invalid auction_id
+        require(auction_id <= storage.auction_count, Error::InvalidAuctionID);
+
+        calculate_price(auction_id)
+    }
+
+    fn bid(auction_id: u64) {
+        /// In a Dutch auction the first bid wins
+        require(storage.initialized, Error::ContractNotYetInitialized);
+
+        /// If the given auction id is higher than the auction count, its an invalid auction_id
+        require(auction_id <= storage.auction_count, Error::InvalidAuctionID);
+
+        let mut auction = storage.auctions.get(auction_id);
+        let price = calculate_price(auction_id);
+
+        /// Cannot bid before auction starts
+        require(height() >= auction.start_time, Error::AuctionNotYetStarted);
+
+        /// Checks for correct asset_id being sent and high enough amount being sent
+        require(msg_asset_id() == auction.asset_id, Error::WrongAssetSent);
+        require(price <= msg_amount(), Error::BidTooLow);
+
+        require(!auction.ended, Error::AuctionAlreadyEnded);
+
+        /// Disallows furthur bids
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+
+        /// If someone sends more than the current price, refunds the extra amount
+        if msg_amount() > price {
+            let return_amount = msg_amount() - price;
+            transfer_to_identity(return_amount, auction.asset_id, get_sender_identity());
+        }
+
+        on_win(auction_id, price);
```

Comment:
since you already have the auction you can pass it in as an argument instead of querying storage again in the function